### PR TITLE
Williamhe/1732771 refactor automation

### PR DESCRIPTION
**Pull request by william-msft**
- [x] The purpose of this PR is explained in this message or in an issue. If an issue please include a reference as:
  - Re-factor DotNet automation which includes
    - Getting all new DotNet versions
      - Before we would match the release date to the current date
      - We also added semantic version range; supporting a minimum version
    - Updated to use dependency injection
    - Separated deserialization classes into their own Models folder
- [x] Tests are included and/or updated for code changes.
- [x] Proper license headers are included in each file.

### CODE REVIEW

File: build/tools/Automation/DotNet/DotNet.cs
Comment (cormacpayne) [inline build/tools/Automation/DotNet/DotNet.cs:49]: We should make `oryxRootPath` a private variable in this class that we set after it's been provided to `RunAsync` so that every method is able to easily reference it without us having to pass it around in each method. This value should only ever be provided once on application startup (since it's an argument to the app), so there shouldn't be any concern about having it overridden or changed during the lifespan of it executing.

File: build/tools/Automation/DotNet/Models/FileObj.cs
Comment (cormacpayne) [inline build/tools/Automation/DotNet/Models/FileObj.cs:9]: nit: I'm typically opposed to including `Obj` in the name of a class since it's already implied that it'll be an object once instantiated, but if this makes more sense for you, then I'm OK keeping it this way

(this comment applies for all class with the `Obj` suffix)

File: build/tools/Automation/DotNet/Models/FileObj.cs
Comment (cormacpayne) [inline build/tools/Automation/DotNet/Models/FileObj.cs:15]: nit: what does `rid` mean in this context?

File: build/tools/Automation/Models/YamlConstants.cs
Comment (cormacpayne) [inline build/tools/Automation/Models/YamlConstants.cs:12]: I would avoid having the `Constants` suffix on this class name if the class isn't encapsulating static string constants (I know this is focused on the `constants.yaml` file, so maybe a better name is `ConstantsYamlFile`)

File: build/tools/Automation/Program.cs
Comment (cormacpayne) [inline build/tools/Automation/Program.cs:45]: nit: we should use `platform` rather than `args[0]` here since it's a variable we already defined

File: build/tools/Automation/Services/VersionService.cs
Comment (cormacpayne) [inline build/tools/Automation/Services/VersionService.cs:47]: Any reason the `exceptionVersions` argument is a string of comma-separated versions and isn't just passed in as an array of versions?

File: build/tools/Automation/Services/YamlFileReaderService.cs
Comment (cormacpayne) [inline build/tools/Automation/Services/YamlFileReaderService.cs:18]: We should have some error handling when dealing with I/O

File: build/tools/Automation/Telemetry/LoggerImpl.cs
Comment (cormacpayne) [inline build/tools/Automation/Telemetry/LoggerImpl.cs:9]: Similar to the `HttpClientImpl` comment -- what is the purpose of this class?

File: tests/AutomationTests/DotNetTests.cs
Comment (cormacpayne) [inline tests/AutomationTests/DotNetTests.cs:1]: Please add a license header

Comment (william-msft) [reply tests/AutomationTests/DotNetTests.cs]: Updated

File: build/tools/Automation/DotNet/Models/FileObj.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/Models/FileObj.cs:9]: So `File` belongs to belongs to `System.IO.File`. I'll just have to specify the namespace to differentiate. What are you thoughts on that?

File: build/tools/Automation/Services/VersionService.cs
Comment (william-msft) [reply build/tools/Automation/Services/VersionService.cs:47]: Originally I was thinking of storing the `exceptionVersions` as an environment variable string, and passing this to this method to avoid having to submit a PR every time we want to skip a version. But also a part of me is also thinking of storing the `exceptionVersions` as an array in a `Constants.cs` file. We'll just have to submit  PR every time we want to skip.

File: build/tools/Automation/Client/HttpClientImpl.cs
Comment (william-msft) [reply build/tools/Automation/Client/HttpClientImpl.cs]: Updated using an extension class and have added documentation. Thank you for helping me with creating this an extension 😄

File: build/tools/Automation/Client/HttpClientImpl.cs
Comment (william-msft) [reply build/tools/Automation/Client/HttpClientImpl.cs]: Updated

File: build/tools/Automation/Services/YamlFileReaderService.cs
Comment (william-msft) [reply build/tools/Automation/Services/YamlFileReaderService.cs]: Updated

File: build/tools/Automation/DotNet/DotNet.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNet.cs:67]: Updated.

File: build/tools/Automation/Services/VersionService.cs
Comment (cormacpayne) [reply build/tools/Automation/Services/VersionService.cs:47]: Independent of how we actually provide the versions-to-skip to this program (I agree it should be an environment variable that we can set in the _Variables_ section of our repository), this question was more about the actual `exceptionVersions` argument to this method being a comma-separated `string` rather than a `string[]` to make it easier to work with.

File: build/tools/Automation/DotNet/Models/FileObj.cs
Comment (cormacpayne) [reply build/tools/Automation/DotNet/Models/FileObj.cs:9]: I think if we set up the classes correctly (_i.e._, if we contain all of the I/O calls within their own `FileService` class), then I would be optimistic we wouldn't use a `File` object in the same class that has a `using System.IO` reference.

If that's not the case, then we could defer to using the `Metadata` suffix instead of `Obj`, which should make sense in the context of these classes you've created.

File: build/tools/Automation/Extensions/HttpClientExtensions.cs
Comment (cormacpayne) [inline build/tools/Automation/Extensions/HttpClientExtensions.cs:39]: nit: let's combine these two lines together

```cs
return await response.Content.ReadAsStringAsync();
```

File: build/tools/Automation/Extensions/HttpClientExtensions.cs
Comment (cormacpayne) [inline build/tools/Automation/Extensions/HttpClientExtensions.cs:68]: We still need to handle the case where `response` is `null` here -- we should be able to do a quick `null` check and return `versions` if it is

File: build/tools/Automation/Program.cs
Comment (cormacpayne) [inline build/tools/Automation/Program.cs:28]: Similar comment in the Python PR, but it would be nice to rename these platform processing classes to something like `DotNetProcessor` or `DotNetAutomator` so that we don't need the redundant namespace suffix and the class name more accurately describes what it's responsible for.

File: tests/AutomationTests/DotNetTests.cs
Comment (cormacpayne) [inline tests/AutomationTests/DotNetTests.cs:5]: Are we planning on adding tests before this PR is merged?

**Review by cormacpayne — COMMENTED** [review]
Some additional comments to take a look at -- feel free to ping me offline if you'd like to sync about any of them

File: build/tools/Automation/Services/VersionService.cs
Comment (william-msft) [reply build/tools/Automation/Services/VersionService.cs:47]: Updated to use List<string>

File: build/tools/Automation/DotNet/Constants.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/Constants.cs]: Updated

File: build/tools/Automation/Program.cs
Comment (william-msft) [reply build/tools/Automation/Program.cs]: Updated

File: tests/AutomationTests/DotNetTests.cs
Comment (william-msft) [reply tests/AutomationTests/DotNetTests.cs]: Yes, working on updating this

File: build/tools/Automation/Extensions/HttpClientExtensions.cs
Comment (william-msft) [reply build/tools/Automation/Extensions/HttpClientExtensions.cs]: Updated

File: build/tools/Automation/DotNet/Constants.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/Constants.cs]: I've added a Constant.cs under Automation/Constants.cs

File: build/tools/Automation/DotNet/DotNet.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNet.cs]: Updated

File: build/tools/Automation/Client/HttpClientImpl.cs
Comment (william-msft) [reply build/tools/Automation/Client/HttpClientImpl.cs]: Updated

File: build/tools/Automation/Telemetry/LoggerImpl.cs
Comment (william-msft) [reply build/tools/Automation/Telemetry/LoggerImpl.cs]: I've converted Logger to an extensions class. I'm thinking of adding the telemetry code in the python PR

File: build/tools/Automation/DotNet/DotNet.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNet.cs]: Updated to use via RunAsync

File: build/tools/Automation/DotNet/DotNet.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNet.cs]: I've updated the the logic a little here so we don't need to check for `null`

File: build/tools/Automation/Automation.csproj
Comment (cormacpayne) [inline build/tools/Automation/Automation.csproj:11]: Are either of these packages used in the `Automation` project?

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (cormacpayne) [inline build/tools/Automation/DotNet/DotNetAutomator.cs:210]: nit: do we need to provide the newline character as a prefix to the second argument in `UpdateVersionsToBuildTxt` or can we automatically set it inside of the method when the argument value is used?

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (cormacpayne) [inline build/tools/Automation/DotNet/DotNetAutomator.cs:241]: I may have missed the previous comment resolution, but is there a reason we can't just always use the `{majorVersion}_{minorVersion}` pattern as a key?

File: build/tools/Automation/Extensions/HttpClientExtensions.cs
Comment (cormacpayne) [inline build/tools/Automation/Extensions/HttpClientExtensions.cs:32]: nit: would you mind removing the `.` character before the `:` character on this line?

```
Response received: {response.StatusCode}
```

File: build/tools/Automation/Program.cs
Comment (cormacpayne) [inline build/tools/Automation/Program.cs:31]: nit: would you make sure that the mappings here are in lexicographical order based?

```cs
.AddSingleton<IFileService>(...)
.AddSingleton<IVersionService, VersionService>()
.AddSingleton<IYamlFileService>(...)
```

File: build/tools/Automation/Services/IYamlFileService.cs
Comment (cormacpayne) [inline build/tools/Automation/Services/IYamlFileService.cs:15]: Let's remove `public` here

**Review by cormacpayne — COMMENTED** [review]
A few additional comments on top of our `IHttpService` discussion, otherwise it LGTM :shipit:

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNetAutomator.cs:203]: Doesn't seem to need it. But I've added an `Environment.NewLine` after the `line` that is provided in the `FileService`

File: build/tools/Automation/Program.cs
Comment (william-msft) [reply build/tools/Automation/Program.cs]: Updated

File: build/tools/Automation/Services/IYamlFileService.cs
Comment (william-msft) [reply build/tools/Automation/Services/IYamlFileService.cs:15]: Ah yes, removed.

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNetAutomator.cs:234]: None that I could see. YAML allows `_` and underscore also is not part of semantic versioning.

Comment (william-msft) [reply build/tools/Automation/DotNet/DotNetAutomator.cs:234]: Side note: I've also updated this for PythonAutomator

File: build/tools/Automation/Extensions/HttpClientExtensions.cs
Comment (william-msft) [reply build/tools/Automation/Extensions/HttpClientExtensions.cs]: Remove

File: build/tools/Automation/Services/IHttpService.cs
Comment (cormacpayne) [inline build/tools/Automation/Services/IHttpService.cs:12]: nit: the `public` access modifier isn't needed for interface methods (prior to C# 8.0, it would've caused a compilation error) since it's the default value when no access modifier is presented, but it's OK to keep this as-is (unless there are other changes that need to be made, then it would be a nice clean up change to make)

File: tests/AutomationTests/Services/YamlFileServiceTests.cs
Comment (cormacpayne) [inline tests/AutomationTests/Services/YamlFileServiceTests.cs:25]: Since we're writing to files in this test suite, we should create a temp directory for the files we're creating so there's no potential crossover of tests targeting the same files or paths. The easiest way to do this would be to generate a random GUID that acts as the intermediate folder between the root path that we're using and any files that we'd be writing to.

File: tests/AutomationTests/Services/YamlFileServiceTests.cs
Comment (cormacpayne) [inline tests/AutomationTests/Services/YamlFileServiceTests.cs:33]: For the tests in this file where we're writing the contents to a new file, we should handle the case where an unexpected error is thrown but we still delete the file we generated. In this case, if the `ReadConstantsYamlFileAsync` method below fails for an unexpected error, we never hit the `File.Delete` call at the bottom and the file we wrote to lingers around (which isn't a _huge_ deal since these run on separate container agents, but locally it may linger).

I suggest wrapping the test with a `try-finally`:

```cs
try
{
    // Arrange
   ...

    // Act
    ...

    // Assert
    ...
}
finally
{
    // Clean up
    File.Delete(this.testConstantsYamlFilePath);
}
```

**Review by cormacpayne — DISMISSED** [review]
A couple of minor comments with the tests, but otherwise LGTM :shipit:

File: build/tools/Automation/Constants.cs
Comment (daniv-msft) [inline build/tools/Automation/Constants.cs:4]: Could you please clarify if there are any reasons to remove the MIT license header here? Looks like it should still be valid in that case.

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (daniv-msft) [inline build/tools/Automation/DotNet/DotNetAutomator.cs:91]: I'm wondering if in the case where `response == null`, we should just throw an exception/exit the flow early.
Could you please clarify if there are any advantages to continue if we didn't get any reply there?

Comment (daniv-msft) [reply build/tools/Automation/DotNet/DotNetAutomator.cs:91]: Also, I'm assuming that this would mostly happen if the .NET team changes the path where they store the release index, right? If that's the case, how would we detect that an issue happened?

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (daniv-msft) [inline build/tools/Automation/DotNet/DotNetAutomator.cs:111]: This would only happen if the .NET team had messed up versions (something like the .NET 7 release index containing .NET 6 releases), right? If it's the case, then please consider throwing an exception as this should never happen, and if it does we should let them know.

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (daniv-msft) [inline build/tools/Automation/DotNet/DotNetAutomator.cs:131]: Nit: please consider removing the re-assignment on `dotNetVersions` as we're about to return anyway.

File: build/tools/Automation/Services/IHttpService.cs
Comment (william-msft) [reply build/tools/Automation/Services/IHttpService.cs:12]: Updated. I might have copied the signature from the `HttpService`

File: tests/AutomationTests/Services/YamlFileServiceTests.cs
Comment (william-msft) [reply tests/AutomationTests/Services/YamlFileServiceTests.cs:25]: Updated to use GUID

File: tests/AutomationTests/Services/YamlFileServiceTests.cs
Comment (william-msft) [reply tests/AutomationTests/Services/YamlFileServiceTests.cs:33]: Great idea! Especially since we only have 10GB disk space in our agents.

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNetAutomator.cs:111]: We need to check our storage account twice because on how .NET releases information is structured. Both release index and releases can have versions that we should check for in our storage account. 

Here is an example flow:
1. GET https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/releases-index.json
a. Check storage account once to make sure the latest version is not already in storage account
b. This is the release index
3. Get releases URL from `"releases.json"`
4. GET https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/7.0/releases.json
a. Check storage account again. Since since the `"releases"` elements has all `7.0` versions

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNetAutomator.cs:131]: Thanks for catching this one!

File: build/tools/Automation/Constants.cs
Comment (william-msft) [reply build/tools/Automation/Constants.cs:4]: Sorry, I accidentally deleted. But have added back

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (william-msft) [reply build/tools/Automation/DotNet/DotNetAutomator.cs:91]: These should never be empty. So it would make sense to throw an  exception for when they are. I've updated the code for this.

File: build/tools/Automation/DotNet/DotNetAutomator.cs
Comment (daniv-msft) [reply build/tools/Automation/DotNet/DotNetAutomator.cs:101]: OK fair enough. I was mostly focusing on validating that the version is correct, but it's true that we also need to check if things already exist.

**Review by daniv-msft — DISMISSED** [review]

**Review by daniv-msft — DISMISSED** [review]

**Review by daniv-msft — DISMISSED** [review]

**Review by daniv-msft — APPROVED** [review]

**Pull request merged**
